TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot/subsystems/XDrive.java

```diff
@@ -0,0 +1,187 @@
+package org.firstinspires.ftc.teamcode.robot.subsystems;
+
+import static org.firstinspires.ftc.teamcode.robot.constants.XDriveConstants.*;
+
+import androidx.annotation.NonNull;
+
+import com.qualcomm.robotcore.hardware.DcMotor;
+import com.qualcomm.robotcore.hardware.HardwareMap;
+
+import org.firstinspires.ftc.robotcore.external.Telemetry;
+import com.qualcomm.hardware.bosch.BNO055IMU;
+import com.qualcomm.robotcore.hardware.IMU;
+import com.qualcomm.robotcore.hardware.ImuOrientationOnRobot;
+import com.qualcomm.hardware.rev.RevHubOrientationOnRobot;
+
+import org.firstinspires.ftc.robotcore.external.navigation.Orientation;
+import org.firstinspires.ftc.robotcore.external.navigation.AxesReference;
+import org.firstinspires.ftc.robotcore.external.navigation.AxesOrder;
+import org.firstinspires.ftc.robotcore.external.navigation.AngleUnit;
+import org.firstinspires.ftc.robotcore.external.navigation.Quaternion;
+import org.firstinspires.ftc.teamcode.core.lib.gamepad.GamepadManager;
+import org.firstinspires.ftc.teamcode.core.lib.gamepad.SmartGamepad;
+import org.firstinspires.ftc.teamcode.core.lib.interfaces.Subsystem;
+import org.firstinspires.ftc.teamcode.core.util.MathUtils;
+
+import java.util.HashMap;
+
+public class XDrive implements Subsystem {
```

Comment:
fy

I believe we have a lot of logic on this XDrive all in one subystem. 
my suggestion is try to break in small parts (like by sensors responsabilities) and utilize in here.
for instance, getHeading, getParameters, gyroAngle...

Reply:
Tuesday I will be at ifsc and I will look into this. The xDrive subsystem have some codes that can be changed, but I would like to do this with the xDrive to test, to avoid commits with errors

Reply:
I talked with Bruno about some ideias that I had in my mind, and I think the subsystem can improve, but I guess It can be done later. I think this can be resolved with an update on the library, creating a class just for the IMU (because all the control hubs have one)